### ISSUE TITLE
AlwaysIncludeUserClaimsInIdToken + WebForms app - catch 22

### ISSUE THREAD

**Issue by cosmoKenney** [IdentityServer/IdentityServer4#4349]
### Question
I have a webforms app that needs a non-null value in `Context.User.Identity.Name`. 
If I turn on `AlwaysIncludeUserClaimsInIdToken` then I get an actual value in `Context.User.Identity.Name`, but then my id token is too large so when I hit the endsession endpoint via `context.Request.GetOwinContext().Authentication.SignOut()` then I get a 404 at the IDP.
Suggestions?

**Comment by leastprivilege** [comment]
Turn `AlwaysIncludeUserClaimsInIdToken` off and use the userinfo endpoint to retrieve the additional claims.

**Comment by cosmoKenney** [comment]
@leastprivilege -- thanks for the fast response.
I had been using the ids3 webforms sample code verbatim in my startup. And at one point I tried to implement Hybrid so I could enable `offline_access`. But with hybrid the access token is not present, so the code in the `SecurityTokenValidated` handler that maps claims from the user info endpoint into a new `ClaimsIdentity` never ran. 
But after going back to implicit, I am getting all the info i need from the user info endpoint.  I ended up refactoring the ~160 aspx pages to get username from the `preferred_user_name` claim instead of  `Context.User.Identity.Name` -- which now holds the user's full name from the `name` claim. 

All this was because I'm having trouble keeping tokens/cookies from expiring. And we get odd behavior after they do.  I.e. refresh after letting your session idle for longer than the token/cookie lifetime and the app crashes instead of redirecting to IDP. Or if the user tries to logout after token/cookie expires, then IDP crashes because I have dependencies in the IDP on `client_id` which I don't get. 

I need a way to keep their tokens/cookies from expiring or to be notified before they do so I can round-trip them to the IDP. I got lazy with the oidc-client.js library which handles the silent refresh for me in my angular apps. So I'm still researching how to update tokens/cookies in the webforms app.
